index.js

```diff
@@ -72,12 +79,13 @@ Choo.prototype.route = function (route, handler) {
   var self = this
   this.router.on(route, function (params) {
     return function () {
-      self.state.params = params
-      self.state.route = route
+      var state = self.state
+      var emit = self.emitter.emit.bind(self.emitter)
+      var cache = self.cache.render.bind(self.cache)
+      state.params = params
+      state.route = route
       var routeTiming = nanotiming("choo.route('" + route + "')")
-      var res = handler(self.state, function (eventName, data) {
-        self.emitter.emit(eventName, data)
-      })
+      var res = handler(state, emit, cache)
```

Comment:
I'm not entirely sure what this code does. I'll have to take a look at the other PR I guess haha. I trust you it works — and fully subscribe to the benefits you're laying out here.

Also been thinking about Next.js style SSR rendering, and think once this patch lands we can work out the details to get some combination of events / class methods in there to make it streamlined. But this is probably first C:

Reply:
Sorry, I could have been more clear and have supplied a proper example. I added it in there now but didn't wanna update the todo example seeing as the main view wouldn't benefit from it in any way. But for documentations' sake, this is what it would look like:

```javascript
module.exports = class MainView extends Component {
  static identity () {
    return 'main-view'
  }

  update () {
    return true
  }

  createElement (state, emit, render) {
    return html`
      <body>
        <section class="todoapp">
          ${render(Header)}
          ${render(Todos)}
          ${render(Footer)}
        </section>
        ${render(Info)}
      </body>
    `
  }
}
```